### DISCUSSION TITLE
dataset.addDataStyle not working for bar charts

### DISCUSSION THREAD

**Discussion by edkennard** [GSI-CS-CO/chart-fx#497] [Q&A]
Hi,

I have a single `DefaultErrorDataSet` added to a bar chart that's using a `HistogramRenderer` configured as follows:

```
val renderer = new HistogramRenderer()
renderer.setDrawBars(true)
renderer.setErrorType(ErrorStyle.NONE)
renderer.setPolyLineStyle(LineStyle.NONE)
renderer.setShiftBar(false)
renderer.setRoundedCorner(false)
```

I want some bars in the dataset to be rendered one colour, and others another colour.  I've tried using `dataset.addDataStyle(index, "strokeColor=#FF4C1E;")` but the chart ignores these settings.

Instead, as a workaround I've had to create two separate datasets, one for each colour, with blank values in each dataset set to 0.0.  I've then applied the different colours using `dataSet.setStyle("strokeColor=#FF4C1E;")`. This creates some undesirable side-effects though, for example with tooltips.

Should `dataset.addDataStyle` work in this scenario, or is my workaround actually the recommended approach?

Thanks!

Ed

**Comment by wirew0rm** [comment]
Hey Ed,

it seems like the Renderer only implements setting the line style (which you disabled) but not the fill:
https://github.com/GSI-CS-CO/chart-fx/blob/master/chartfx-chart/src/main/java/de/gsi/chart/renderer/spi/HistogramRenderer.java#L270
If you look into the implementation of `DefaultRendererColorScheme::setLineScheme` you can see that the fill is only set based on the dataset index and the point style has no effect.
This is because the `DefaultRendererColorScheme` is shared with other renderers, which use different routines for drawing the lines and the surfaces.
Please feel free to add a PR for adding this or convert this discussion to an appropriate issue so we can fix this ourself.

As for workarounds, you could switch to use outline bars instead of filled bars. Or just override the renderer method to also call `DefaultRendererColorScheme::setFill` until this is fixed in ChartFx.

**Reply by edkennard** [reply]
@wirew0rm Thanks so much for your swift reply, it's a really positive point when working with your library.  I was able to join our daily team meeting this morning armed with a clear path to completion on my work integrating chart-fx into our project :)

I will submit a PR to you with a fix which I've already tested and is working well. I added a line to `HistogramRenderer` immediately after the one you highlighted previously, as follows:

```java
if (dataPointStyle != null) {
  gc.save();
  DefaultRenderColorScheme.setMarkerScheme(gc, dataPointStyle, lindex);
  DefaultRenderColorScheme.setLineScheme(gc, dataPointStyle, lindex);
  DefaultRenderColorScheme.setFillScheme(gc, dataPointStyle, lindex);              //   <-- Added this line
  DefaultRenderColorScheme.setGraphicsContextAttributes(gc, dataPointStyle);
}
```

The calling code now does this:
```java
dataset.addDataStyle(index, "fillColor=#FF4C1E;")
```

The original tests all pass, but I will update them to reflect this change before submitting the PR

Cheers

Ed

**Comment by edkennard** [comment]
The fix for this has now been submitted - #500
